### ISSUE TITLE
Makefile broken on master

### ISSUE THREAD

**Issue by lanzafame** [filecoin-project/lotus#7165]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] This is **not** a question or a support request. If you have any lotus related questions, please ask in the [lotus forum](https://github.com/filecoin-project/lotus/discussions).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** an enhancement request. If it is, please file a [improvement suggestion](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fenhancement&template=enhancement.yml) instead.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [X] I did not make any code changes to lotus.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [ ] lotus miner/market - storage deal
- [ ] lotus miner/market - retrieval deal
- [ ] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [X] Other

### Lotus Version

```text
master
```

### Describe the Bug

This issue may just be with the build script in the rust libs that lotus imports, but I am not sure of the exact mechanics. 

This here is the offending line:
```
+ cargo +nightly-2021-04-24 lipo --release --targets x86_64-apple-darwin,aarch64-apple-darwin --no-default-features --features multicore-sdr,blst,gpu
```

Error:
```
error: no such subcommand: `lipo`

        Did you mean `fix`?
```

### Logging Information

```text
$ env RUSTFLAGS="-C target-cpu=native -g" FFI_BUILD_FROM_SOURCE=1 make clean deps
rm -rf  build/.filecoin-install build/.update-modules  lotus lotus-miner lotus-worker lotus-shed lotus-gateway lotus-seed lotus-pond lotus-fountain lotus-bench lotus-stats lotus-pcr lotus-health lotus-wallet lotus-keygen testground tvx lotus-sim
/Library/Developer/CommandLineTools/usr/bin/make -C extern/filecoin-ffi/ clean
go clean -cache -testcache .
rm -rf filcrypto.h filcrypto.pc libfilcrypto.a .install-filcrypto
rm -f ./runner
cd rust && cargo clean && cd ..
/Users/lanzafame/src/github.com/filecoin-project/lotus/extern/filecoin-ffi/rust
git submodule update --init --recursive
touch build/.update-modules
/Library/Developer/CommandLineTools/usr/bin/make -C extern/filecoin-ffi/ .install-filcrypto
go clean -cache -testcache .
./install-filcrypto
+ auth_header=()
+ '[' -n 9aa9cea2caea7f3e633fd20ce08d9118d4700d26 ']'
+ auth_header=("-H" "Authorization: token ${GITHUB_TOKEN}")
++ dirname ./install-filcrypto
+ cd .
+ rust_sources_dir=rust
++ jq -r '.[].check_cpu_for_feature | select(. != null)'
+ cpu_features_required_for_optimized_release='adx
sha_ni
sha2
sse2
avx2
avx
sse4_2
sse4_1'
+ main
++ get_release_flags
++ local __features=
++ [[ ! -f /proc/cpuinfo ]]
++ echo '[get_release_flags] no /proc/cpuinfo file; falling back to Darwin feature detection'
[get_release_flags] no /proc/cpuinfo file; falling back to Darwin feature detection
+++ sysctl -a
+++ grep machdep.cpu
+++ tr '[:upper:]' '[:lower:]'
+++ grep features
++ __features='machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ uname -s
++ '[' Darwin = Darwin ']'
+++ uname -m
++ '[' x86_64 = x86_64 ']'
++ feature_map=("adx:+adx" "sha_ni:+sha" "sha2:+sha2" "avx2:+avx2" "sse4_2:+sse4.2" "sse4_1:+sse4.1")
++ target_features=
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
+++ grep -c adx
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
++ current_feature=1
++ '[' 1 = 1 ']'
++ for feature in '"${feature_map[@]}"'
++ key=adx
++ '[' adx == adx ']'
++ val=+adx
++ '[' -z '' ']'
++ target_features=+adx
++ for feature in '"${feature_map[@]}"'
++ key=sha_ni
++ '[' sha_ni == adx ']'
++ for feature in '"${feature_map[@]}"'
++ key=sha2
++ '[' sha2 == adx ']'
++ for feature in '"${feature_map[@]}"'
++ key=avx2
++ '[' avx2 == adx ']'
++ for feature in '"${feature_map[@]}"'
++ key=sse4_2
++ '[' sse4_2 == adx ']'
++ for feature in '"${feature_map[@]}"'
++ key=sse4_1
++ '[' sse4_1 == adx ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c sha_ni
++ current_feature=0
++ '[' 1 = 0 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c sha2
++ current_feature=0
++ '[' 1 = 0 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c sse2
++ current_feature=1
++ '[' 1 = 1 ']'
++ for feature in '"${feature_map[@]}"'
++ key=adx
++ '[' adx == sse2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sha_ni
++ '[' sha_ni == sse2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sha2
++ '[' sha2 == sse2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=avx2
++ '[' avx2 == sse2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sse4_2
++ '[' sse4_2 == sse2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sse4_1
++ '[' sse4_1 == sse2 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c avx2
++ current_feature=1
++ '[' 1 = 1 ']'
++ for feature in '"${feature_map[@]}"'
++ key=adx
++ '[' adx == avx2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sha_ni
++ '[' sha_ni == avx2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sha2
++ '[' sha2 == avx2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=avx2
++ '[' avx2 == avx2 ']'
++ val=+avx2
++ '[' -z +adx ']'
++ target_features=+adx,+avx2
++ for feature in '"${feature_map[@]}"'
++ key=sse4_2
++ '[' sse4_2 == avx2 ']'
++ for feature in '"${feature_map[@]}"'
++ key=sse4_1
++ '[' sse4_1 == avx2 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c avx
++ current_feature=2
++ '[' 1 = 2 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c sse4_2
++ current_feature=0
++ '[' 1 = 0 ']'
++ for x in '${cpu_features_required_for_optimized_release[@]}'
+++ echo 'machdep.cpu.extfeatures: syscall xd 1gbpage em64t lahf lzcnt prefetchw rdtscp tsci
machdep.cpu.features: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clfsh ds acpi mmx fxsr sse sse2 ss htt tm pbe sse3 pclmulqdq dtes64 mon dscpl vmx est tm2 ssse3 fma cx16 tpr pdcm sse4.1 sse4.2 x2apic movbe popcnt aes pcid xsave osxsave seglim64 tsctmr avx1.0 rdrand f16c
machdep.cpu.leaf7_features: rdwrfsgs tsc_thread_offset sgx bmi1 avx2 smep bmi2 erms invpcid fpu_csds mpx rdseed adx smap clfsopt ipt sgxlc mdclear tsxfa ibrs stibp l1df ssbd'
+++ grep -c sse4_1
++ current_feature=0
++ '[' 1 = 0 ']'
++ echo +adx,+avx2
+ local __release_flags=+adx,+avx2
+ '[' 1 '!=' 1 ']'
+ echo '[install-filcrypto/main] building libfilcrypto from local sources (dir = rust)'
[install-filcrypto/main] building libfilcrypto from local sources (dir = rust)
+ build_from_source filcrypto rust +adx,+avx2
+ local __library_name=filcrypto
+ local __rust_sources_path=rust
+ local __release_flags=+adx,+avx2
++ git rev-parse HEAD
+ local __repo_sha1=6e901ee3dd5399f5aa3395d48f7d9540b661b944
+ local __repo_sha1_truncated=6e901ee3dd5399f5
+ echo 'building from source @ 6e901ee3dd5399f5'
building from source @ 6e901ee3dd5399f5
++ command -v cargo
+ '[' -x /Users/lanzafame/.cargo/bin/cargo ']'
++ command -v rustup
+ '[' -x /Users/lanzafame/.cargo/bin/rustup ']'
+ pushd rust
/Users/lanzafame/src/github.com/filecoin-project/lotus/extern/filecoin-ffi/rust
~/src/github.com/filecoin-project/lotus/extern/filecoin-ffi/rust ~/src/github.com/filecoin-project/lotus/extern/filecoin-ffi
+ cargo --version
cargo 1.53.0-nightly (65d57e6f3 2021-04-04)
+ additional_flags=
++ uname -s
+ '[' Darwin = Darwin ']'
++ uname -m
+ '[' x86_64 = x86_64 ']'
+ build=lipo
+ additional_flags='--targets x86_64-apple-darwin,aarch64-apple-darwin '
+ gpu_flags=,gpu
+ '[' '' == 0 ']'
+ use_multicore_sdr=multicore-sdr
+ '[' '' == 0 ']'
+ '[' '' == 1 ']'
+ '[' '' == 1 ']'
+ '[' '' == 1 ']'
+ additional_flags='--targets x86_64-apple-darwin,aarch64-apple-darwin  --no-default-features --features multicore-sdr,blst,gpu'
+ echo 'Using additional build flags: --targets x86_64-apple-darwin,aarch64-apple-darwin  --no-default-features --features multicore-sdr,blst,gpu'
Using additional build flags: --targets x86_64-apple-darwin,aarch64-apple-darwin  --no-default-features --features multicore-sdr,blst,gpu
+ '[' -n +adx,+avx2 ']'
++ cat rust-toolchain
+ RUSTFLAGS='-C target-feature=+adx,+avx2'
+ ./scripts/build-release.sh filcrypto nightly-2021-04-24 lipo '--targets x86_64-apple-darwin,aarch64-apple-darwin  --no-default-features --features multicore-sdr,blst,gpu'
+ main filcrypto nightly-2021-04-24 lipo '--targets x86_64-apple-darwin,aarch64-apple-darwin  --no-default-features --features multicore-sdr,blst,gpu'
+ [[ -z filcrypto ]]
+ [[ -z nightly-2021-04-24 ]]
+ [[ -z lipo ]]
++ mktemp
+ __build_output_log_tmp=/var/folders/8d/gjm5s5rd68jdzftb3wvhshpm0000gn/T/tmp.iAGmsYfQ
+ trap '{ rm -f $__build_output_log_tmp; }' EXIT
+ local '__rust_flags=--print native-static-libs -C target-feature=+adx,+avx2'
+ RUSTFLAGS='--print native-static-libs -C target-feature=+adx,+avx2'
+ cargo +nightly-2021-04-24 lipo --release --targets x86_64-apple-darwin,aarch64-apple-darwin --no-default-features --features multicore-sdr,blst,gpu
+ tee /var/folders/8d/gjm5s5rd68jdzftb3wvhshpm0000gn/T/tmp.iAGmsYfQ
error: no such subcommand: `lipo`

        Did you mean `fix`?
+ rm -f /var/folders/8d/gjm5s5rd68jdzftb3wvhshpm0000gn/T/tmp.iAGmsYfQ
make[1]: *** [.install-filcrypto] Error 101
make: *** [build/.filecoin-install] Error 2
```
```


### Repo Steps

1. Run '...'
2. Do '...'
3. See error '...'
...

**Comment by travisperson** [comment]
Work around for this I believe is to manually install https://lib.rs/crates/cargo-lipo on your system.

@cryptonemo should this be added as a requirement for building on macOS in documentation or is this something the filecoin-ffi should handle as a dependency? Kind of feels like documentation is the right place for this but unsure if this is easily handled in the Cargo.toml or whatever.

**Comment by lanzafame** [comment]
Cheers. My preference would be for it to be downloaded as part of the Makefile setup. This isn't a system package but a cargo dependency, especially given that it shouldn't be required for macOS-Intel builds given that platform doesn't support iOS targets.

**Comment by cryptonemo** [comment]
Right, at one point the install file was pulling it in, thought that was still there, but perhaps not anymore.  I'm ok with either -- docs or just installing before using.  Preference?

**Comment by travisperson** [comment]
Having lipo managed during the `build_from_source` step would probably be a good solution I think. My only concern is that running the `cargo install` would result in the tool being installed for the user. This isn't the best side effect.

It seem like there is a way around this, but I'm unsure of the impact it would have on the rest of the build process.

It cargo has [a few ways to manage](https://doc.rust-lang.org/cargo/commands/cargo-install.html#description) the destination during `install`. It might be a good idea to have any additional tooling require for the installation process to be placed at a local location. This will also make it easy to lock to a particular version to avoid issues later down the road.

```
PATH=$(pwd)/cargo_install/bin
CARGO_INSTALL_ROOT=$(pwd)/cargo_install
```

**Comment by cryptonemo** [comment]
I noticed the `./build.sh` script does the install (I guess that's what I was thinking of).  The install script does have the option of installing rust to a specific location (if not already present) and then possibly removing it after, which I had proposed as an alternative to pre-built binaries in the past.  Not much excitement generated there.  OS X seems to be such a niche platform, manually installing cargo-lipo may not be a big deal (not sure).
